### PR TITLE
parallel cache purge

### PR DESCRIPTION
**Pull request by fotanus**

### CODE REVIEW

**Comment by andremedeiros** [comment]
Shouldn't we `join` these threads to make sure everything was purged correctly?

**Comment by fotanus** [comment]
Yeah, would be great if exceptions are raised :+1:
Not sure where to join? Suggestion pushed.
Ideally would be after the response to the client, right?

File: lib/bundler_api/cdn.rb [hunk 1]
```diff
@@ -3,17 +3,21 @@
...
-    client.purge_key  'dependencies'
-    client.purge_path '/latest_specs.4.8.gz'
-    client.purge_path '/specs.4.8.gz'
-    client.purge_path '/prerelease_specs.4.8.gz'
+    threads = []
+    threads << client.purge_key('dependencies')
+    threads << client.purge_path( '/latest_specs.4.8.gz')
+    threads << client.purge_path( '/specs.4.8.gz')
+    threads << client.purge_path( '/prerelease_specs.4.8.gz')
+    threads.each { |t| t.join }
```
Comment (andremedeiros) [inline lib/bundler_api/cdn.rb]: That's better, although I would create the thread at this level and not in `purge_key` or `purge_path`.

Something like:

```ruby
threads = []
threads << Thread.new { client.purge_key('dependencies') }
threads << Thread.new { client.purge_path( '/latest_specs.4.8.gz') }
# ...
threads.each(&:join)
```

**Comment by fotanus** [comment]
This conflicts with #81, so better close for now. I will send another patch later.

**Comment by fotanus** [comment]
I just closed the wrong pull request :-)

**Comment by indirect** [comment]
Maybe I'm being to paranoid here, but is the `client` object threadsafe? If it has internal state, this will likely not work...

**Comment by andremedeiros** [comment]
It doesn't seem to me that any of these operations change the client's state. An HTTP object is [built per request][1], and a client instance gets [instanciated per request][2]. I'm fairly certain that we're fine.

[1]: https://github.com/fotanus/bundler-api/blob/8d28550e0420058e451ce22134cacb81ef833928/lib/bundler_api/cdn.rb#L56-L61
[2]: https://github.com/fotanus/bundler-api/blob/8d28550e0420058e451ce22134cacb81ef833928/lib/bundler_api/cdn.rb#L28-L31

**Comment by indirect** [comment]
Cool, sounds good, then.

**Comment by segiddins** [comment]
:+1:

**Pull request merged**